p2p/src/peer_manager/peers_eviction.rs

```diff
@@ -0,0 +1,114 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeMap, hash::Hasher};
+
+use crypto::random::Rng;
+
+use crate::{
+    net::{default_backend::transport::TransportAddress, types::Role},
+    peer_manager::address_groups::get_address_group,
+    types::peer_id::PeerId,
+};
+
+use super::peer_context::PeerContext;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+struct NetGroupKeyed(u64);
+
+pub struct EvictionCandidate {
+    peer_id: PeerId,
+
+    /// Deterministically randomized address group ID
+    net_group_keyed: NetGroupKeyed,
+
+    /// Minimum ping time in microseconds (or i64::MAX if not known yet)
+    ping_min: i64,
+
+    /// Inbound or Outbound
+    role: Role,
+}
+
+pub struct RandomState(u64, u64);
+
+impl RandomState {
+    pub fn new<R: Rng>(rng: &mut R) -> Self {
+        Self(rng.gen(), rng.gen())
+    }
+
+    fn get_hash<A: std::hash::Hash>(&self, value: &A) -> u64 {
+        let mut hasher = siphasher::sip::SipHasher::new_with_keys(self.0, self.1);
+        value.hash(&mut hasher);
+        hasher.finish()
+    }
+}
+
+impl EvictionCandidate {
+    pub fn new<A: TransportAddress>(peer: &PeerContext<A>, random_state: &RandomState) -> Self {
+        EvictionCandidate {
+            peer_id: peer.info.peer_id,
+            net_group_keyed: NetGroupKeyed(
+                random_state.get_hash(&get_address_group(&peer.address.as_peer_address())),
+            ),
+            ping_min: peer.ping_min.map_or(i64::MAX, |val| val.as_micros() as i64),
+            role: peer.role,
+        }
+    }
+}
+
+/// Based on `SelectNodeToEvict` from Bitcoin Core:
+///
+/// Select an inbound peer to evict after filtering out (protecting) peers having
+/// distinct, difficult-to-forge characteristics. The protection logic picks out
+/// fixed numbers of desirable peers per various criteria.
+/// ...
+/// If any eviction candidates remain, the selection logic chooses a peer to evict.
+pub fn select_for_eviction(mut candidates: Vec<EvictionCandidate>) -> Option<PeerId> {
+    // TODO: Do not evict connections from whitelisted IPs
+    candidates.retain(|peer| peer.role == Role::Inbound);
+
+    // Deterministically select 4 peers to protect by netgroup.
+    // An attacker cannot predict which netgroups will be protected
+    candidates.sort_unstable_by_key(|peer| peer.net_group_keyed);
+    candidates.truncate(candidates.len().saturating_sub(4));
+
+    // Protect the 8 nodes with the lowest minimum ping time.
+    // An attacker cannot manipulate this metric without physically moving nodes closer to the target.
+    candidates.sort_unstable_by_key(|peer| -peer.ping_min);
+    candidates.truncate(candidates.len().saturating_sub(8));
+
+    // TODO: Protect 4 nodes that most recently sent us novel transactions accepted into our mempool.
+    // TODO: Protect up to 8 peers that have sent us novel blocks.
```

Comment:
Same here... preserve... probably.

Reply:
Sure, updated (it was copied from Bitcoin Core).